### PR TITLE
Introduce `Waitlist` type for `Messenger`

### PR DESCRIPTION
**Pull request by breathx**
Major changes:
- Waitlist introduced for `Messenger` trait.
- New types introduced.
- Docs provided.
- `IterableDoubleMap` renamed to `IterableByKeyMap`.

Minor changes:
- Replaced H256 with proper types, if available.
- Usage pallet kept this previous logic for now #969.
- Added script for build gear-test crate.

### CODE REVIEW

File: pallets/usage/src/offchain.rs [hunk 1]
```diff
@@ -142,43 +140,40 @@ where
...
-            let prog_id = H256::from_slice(&key[0..32]);
-            let msg_id = H256::from_slice(&key[34..]);
-            Ok((prog_id, msg_id))
-        });
+
+        // We go through storage, finding last key, if present.
+        // Last key was already processed, so we skip it and
+        // start processing from the next one.
+        //
+        // Starting from the begging for absence of last key.
```
Comment (gshep) [inline pallets/usage/src/offchain.rs:148]: misprint: from the beginning

File: common/src/storage/complex/messenger.rs [hunk 2]
```diff
@@ -102,11 +116,33 @@ pub trait Messenger {
...
 
-    // Soon ...
-    // /// Gear waitlist.
-    // type Waitlist;
+    /// Gear waitlist.
+    ///
+    /// Waitlist contains messages, which execution should be
+    /// delayed for some logic.
+    ///
+    /// Message can be inserted into mailbox only in two cases:
```
Comment (gshep) [inline common/src/storage/complex/messenger.rs:126]: misprint? `... into waitlist`

File: common/src/storage/complex/messenger.rs [hunk 3]
```diff
@@ -102,11 +116,33 @@ pub trait Messenger {
...
+    /// 1. Destination program called `gr_wait` while was executing
+    /// this message, so the only this program can remove and
```
Comment (gshep) [inline common/src/storage/complex/messenger.rs:128]: `... so only this program ...` maybe

File: common/src/storage/complex/messenger.rs [hunk 4]
```diff
@@ -102,11 +116,33 @@ pub trait Messenger {
...
+    /// requeue it by `gr_wake` call in any execution.
+    /// 2. The message sent to program, that hadn't finished it's
```
Comment (gshep) [inline common/src/storage/complex/messenger.rs:130]: `its initialization`

**Review by gshep — APPROVED** [review]
Perfect! :+1:  I could find only several tiny misprints.

**Review by SabaunT — APPROVED** [review]
great work! :+1:

File: common/src/storage/complex/messenger.rs [hunk 5]
```diff
@@ -102,11 +116,33 @@ pub trait Messenger {
...
+    /// result of initialization would be available.
+    ///
+    /// Gear runtime also charges rent for holding in waitlist.
+    /// For details, see `pallet-usage`.
+    type Waitlist: Waitlist<
+            Key1 = Self::WaitlistFirstKey,
+            Key2 = Self::WaitlistSecondKey,
+            Value = Self::WaitlistedMessage,
+            BlockNumber = Self::BlockNumber,
+            Error = Self::Error,
```
[the remark this answers is not in the document]
Comment (breathx) [reply common/src/storage/complex/messenger.rs:150]: Yeah, you are right. It doesn't fail due to proper implementation on `gear-messenger-pallet`, but it's important for clarification. Added.

File: common/src/storage/complex/waitlist.rs [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+    /// which uses this callbacks store.
+    type Value;
+    /// Callback relative type.
+    ///
+    /// This type represents block number of stored component in waitlist,
+    /// which uses this callbacks store.
+    type BlockNumber;
+
+    /// Callback used for getting current block number when it need.
+    type OnBlockNumberNeed: GetCallback<Self::BlockNumber>;
```
[the remark this answers is not in the document]
Comment (breathx) [reply common/src/storage/complex/waitlist.rs]: Replaced with `type GetBlockNumber`, seems well: `Callbacks::GetBlockNumber::call()`

File: common/src/storage/complex/messenger.rs [hunk 7]
```diff
@@ -102,11 +116,33 @@ pub trait Messenger {
...
+    /// Waitlist contains messages, which execution should be
+    /// delayed for some logic.
+    ///
+    /// Message can be inserted into waitlist only in two cases:
+    /// 1. Destination program called `gr_wait` while was executing
+    /// this message, so only this program can remove and
+    /// requeue it by `gr_wake` call in any execution.
+    /// 2. The message sent to program, that hadn't finished its
+    /// initialization, and will be automatically removed once
+    /// result of initialization would be available.
```
[the remark this answers is not in the document]
Comment (breathx) [reply common/src/storage/complex/messenger.rs:132]: Yeah, added explanation about resuming and pausing with waitlist recovering. Added text about future possibilities, and added doc about that pallet-usage can remove messages from wl also. Re-check these lines, please

**Pull request merged**